adapter/src/main/java/com/twilio/kudu/sql/KuduEnumerable.java

```diff
@@ -495,10 +495,17 @@ public Schema getTableSchema() {
       if (!columnIndices.isEmpty()) {
         tokenBuilder.setProjectedColumnIndexes(columnIndices);
       }
-      // we can only push down the limit if we are ordering by the pk columns
-      // and if there is no offset
-      if (sort && offset == -1 && limit != -1 && !groupBySorted) {
-        tokenBuilder.limit(limit);
+      // Push down the limit if present AND
+      // 1. Not doing a group aggregation.
+      // 2. All the predicates are pushed into the scan.
+      // When those are true, the Scanners can inform the datanode that it only
+      // requires a small number of rows.
+      if (limit > 0 && !groupBySorted && filterFunction == Predicate1.TRUE) {
```

Comment:
This won't work. `Predicate1.TRUE` is not used in the actual code generation:
```java
// This is builds a predicate function that will always be present. It checks if
    // the RowResult
    // should be returned in the KuduEnumerable.
    final BlockBuilder filterBuilder = new BlockBuilder();
    final Expression condition;
    if (!kuduImplementor.filterProjections.isEmpty()) {
      condition = RexToLixTranslator.translateCondition(projectionFunctions, implementor.getTypeFactory(),
          filterBuilder, inputGetter, null, implementor.getConformance());

    } else {
      condition = RexImpTable.TRUE_EXPR;
    }

    filterBuilder.add(Expressions.return_(null, condition));
    final Expression filterFunction = Expressions.new_(Predicate1.class, Collections.emptyList(),
        Expressions.methodDecl(Modifier.PUBLIC, boolean.class, "apply", Collections.singletonList(inputRow),
            filterBuilder.toBlock()));
```

https://github.com/twilio/calcite-kudu/blob/be5e00ed82c6907a56389792fb8ab09bd5f338f1/adapter/src/main/java/com/twilio/kudu/sql/rel/KuduToEnumerableRel.java#L196-L212

Reply:
Fixed.